thijsvanloef/palworld-server-docker
# Use Supervisord to manage processes

## Context 

* This will enable us to start, stop, restart different processes
* No Container reboot is needed to restart the main Palworld process
* Enables updates, restores and other scripts can be run within the container without restarting it.
* closes #399 

## Choices

* Using Supervisord to manage the processes.
* Adds Triggers for Discord Notifications tied to the application state of the palworld server
* Adds a separate script for starting the server, to only run start.sh once on container start.
* Update and reboot scripts use supervisord to shutdown and restart the server
* Restore Script uses supervisord to shutdown and restart the server


### Processes Managed by Supervisord
* Palworld
* Palworld-Event-Handler
* Supercronic

## Test instructions

### Environment Variables
These Environment variables can be used for all test cases

```env
DISCORD_WEBHOOK_URL: "your test webhook"
UPDATE_ON_BOOT: false
BACKUP_ENABLED: false
PLAYER_LOGGING_POLL_PERIOD: 10
```

### Always Running Services
These services should always be running if enabled in the config
* Palworld-Server
* Crontab
* Player Logging (can be disabled)

#### Test
1. Start Container
2. Verify that the Server logs correctly to docker logs
3. Verify that the discord notifications are being sent correctly
4. Check if all mentioned above are processes are running `docker exec -it palworld-server supervisorctl status`
5. Join the server with an palworld client
6. Verify that the player_logging scripts sends a message as broadcast and in the logs
7. Stop the container
8. Verify with docker logs in another terminal that the server stops correctly
9. Verify that the discord notifications are being sent correctly

### Scripts
These scripts use supervisord to shutdown the server if necessary.
* Restore
* Update
* Auto Reboot

#### Restore Test
1. Start the container
2. Create a backup
3. restore the backup with the command `docker exec -it palworld-server restore`
4. Check the logs and verify that the server shutdown and the restore is successfull

#### Update Test
1. Start the container
2. Run the command `docker exec palworld-server update` and verify that the server is up to date
3. Change the "manifest" value in palworld\steamapps\appmanifest_2394010.acf on Line 31 to a different value
4. Run the command `docker exec palworld-server update` and verify that an update is found
5. Verify that the server is being shutdown without stopping the container
6. Verify that the server is started again after steam has completed the "update"

#### Auto Reboot Test
1. Start the container
2. Run the command `docker exec palworld-server auto_reboot.sh`
3. Verify that the server shutsdown without stopping the container
4. Verify that the server starts again

## Checklist before requesting a review
